File: benchmarks/src/lua/lzw_compress.lua
local M=400000;local seed=12345;local function ri(m)seed=(seed*16807)%2147483647;return seed%m end
local syms={};local i=0
while i<M do if i>=20 and ri(10)<4 then local s=i-20;for j=1,20 do syms[#syms+1]=syms[s+j]end;i=i+20 else syms[#syms+1]=ri(4);i=i+1 end end
local DS=4096;local dic={};for j=1,DS*4 do dic[j]=0 end
local nc=5;local out={};local cur=syms[1]+1;i=2;local start=os.clock()
while i<=#syms do local s=syms[i];local cand=dic[cur*4+s+1];if cand~=0 then cur=cand else out[#out+1]=cur
 if nc<DS then dic[cur*4+s+1]=nc;nc=nc+1 end;cur=s+1 end;i=i+1 end
out[#out+1]=cur;local oc=#out;local sm=0;for j=1,oc do sm=(sm+out[j]*(j-1))%1000003 end
local finish=os.clock();print("Result: "..oc.."_"..sm);print(string.format("Time: %.0fms",(finish-start)*1000))

File: benchmarks/src/perl/lzw_compress.pl
use Time::HiRes qw(time);
my$M=400000;my$seed=12345;my$ri=sub{my$m=shift;$seed=($seed*16807)%2147483647;$seed%$m};
my@syms;my$i=0;
while($i<$M){if($i>=20&&$ri->(10)<4){my$s=$i-20;for(1..20){push@syms,$syms[$s+$_-1]};$i+=20}else{push@syms,$ri->(4);$i++}}
my$DS=4096;my@dic=(0)x($DS*4);my$nc=5;my@out;my$cur=$syms[0]+1;$i=1;my$start=time();
while($i<@syms){my$s=$syms[$i];my$cand=$dic[$cur*4+$s];if($cand!=0){$cur=$cand}else{push@out,$cur;
 if($nc<$DS){$dic[$cur*4+$s]=$nc;$nc++}$cur=$s+1}$i++}
push@out,$cur;my$oc=@out;my$sm=0;for my$j(0..$oc-1){$sm=($sm+$out[$j]*$j)%1000003}
my$end=time();my$ms=($end-$start)*1000;print"Result: ${oc}_${sm}\n";printf"Time: %.0fms\n",$ms;